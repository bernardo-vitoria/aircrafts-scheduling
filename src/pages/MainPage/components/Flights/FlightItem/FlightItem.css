.card-body {
    display: flex;
    justify-content: space-between;
}

.card-body_column {
    display: flex;
    flex-direction: column;
}

.flight-item:hover {
    background-color: red;
}