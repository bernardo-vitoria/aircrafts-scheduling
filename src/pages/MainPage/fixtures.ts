const aircraft = { ident: "GABCD", type: "A320", economySeats: 186, base: "EGKK" };

const aircrafts = [
  { ident: "GABCD", type: "A320", economySeats: 186, base: "EGKK" },
];

const flight = {
  id: "AS1001",
  departuretime: 21600,
  arrivaltime: 26100,
  readable_departure: "06:00",
  readable_arrival: "07:15",
  origin: "LFSB",
  destination: "LFMN",
};

const flights = [
  {
    id: "AS1001",
    departuretime: 21600,
    arrivaltime: 26100,
    readable_departure: "06:00",
    readable_arrival: "07:15",
    origin: "LFSB",
    destination: "LFMN",
  },
  {
    id: "AS1002",
    departuretime: 27900,
    arrivaltime: 32100,
    readable_departure: "07:45",
    readable_arrival: "08:55",
    origin: "LFMN",
    destination: "LFSB",
  },
  {
    id: "AS1025",
    departuretime: 22800,
    arrivaltime: 28200,
    readable_departure: "06:20",
    readable_arrival: "07:50",
    origin: "LFSB",
    destination: "EDDH",
  },
  {
    id: "AS1026",
    departuretime: 30000,
    arrivaltime: 35100,
    readable_departure: "08:20",
    readable_arrival: "09:45",
    origin: "EDDH",
    destination: "LFSB",
  },
];

export {
    aircraft,
    aircrafts,
    flight,
    flights
};
