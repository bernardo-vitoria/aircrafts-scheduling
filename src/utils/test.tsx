// @ts-nocheck
import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { I18nextProvider } from "react-i18next";
import { QueryClient, QueryClientProvider } from "react-query";
import { MemoryRouter } from "react-router-dom"; // 4.0.0
import { queries } from "@testing-library/react";
import {
  render,
  RenderResult,
} from "@testing-library/react";
import i18n from "../config/i18nForTests";

const customRender = (ui: React.ReactElement, options = {}): RenderResult => {
  const CombinedProviders: React.FC = ({ children }) => {
    const config = {
      queries: {
        retry: false,
        cacheTime: 0,
        refetchOnWindowFocus: false,
        retryDelay: 0,
      },
    };

    const queryClient = new QueryClient({ defaultOptions: config });

    return (
        <QueryClientProvider client={queryClient}>
          <I18nextProvider i18n={i18n}>
            <MemoryRouter>{children}</MemoryRouter>
          </I18nextProvider>
        </QueryClientProvider>
    );
  };

  return render(ui, { wrapper: CombinedProviders, queries, ...options });
};

export {
  customRender as render
};