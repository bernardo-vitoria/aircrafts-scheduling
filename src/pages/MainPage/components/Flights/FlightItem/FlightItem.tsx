import React from 'react';
import { styled } from "@mui/material/styles";
import { Paper } from '../../../../../components';
import { useRotationStack } from '../../../../../store';
import { Flight } from "../../../types";
import "./FlightItem.css";

const Item = styled(Paper)(({ theme }) => ({
  ...theme.typography.body2,
  padding: theme.spacing(1),
  textAlign: "center",
  color: theme.palette.text.secondary,
  minWidth: "180px",
  margin: "8px",
  ":hover": { backgroundColor: "#FDFAF6" },
}));

const canAddFlight = (flights: Flight[], value:Flight) => {
  const lastFlightInserted = flights[flights.length - 1];

  if (flights.length === 0) return true;

  const diffBetweenFlights =
    Math.abs((lastFlightInserted.arrivaltime - value.departuretime) * 60);

  if (diffBetweenFlights < 1200) {
    return false;
  }

  if ((value.arrivaltime > 86400000)) {
    return false;
  }

  return true;
};


type Props = {
  flight: Flight;
  hasError: boolean;
  setHasError: (value: boolean) => void;
};

const FlightItem: React.FC<Props> = ({ flight, hasError, setHasError }) => {
  const { id, origin, destination, readable_arrival, readable_departure } = flight;
  const { addFlightToRotation, flightsRotation } = useRotationStack((state) => ({
    addFlightToRotation: state.addFlightToRotation,
    flightsRotation: state.flightsRotation
  }));

  const handleOnClick = () => {
    const canAdd = canAddFlight(flightsRotation,flight);
    setHasError(canAdd);

    if(canAdd) {
      addFlightToRotation(flight); setHasError(false)
    }
    else{
      setHasError(true)
    }
  };

  return (
    <Item onClick={handleOnClick}>
      <span className="flight-item">
        <div>
          <b>{id}</b>
        </div>
        <div className="card-body">
          <div className="card-body_column">
            <div>{origin}</div>
            <div>{readable_departure}</div>
          </div>
          <div className="card-body_column">
            <div>{destination}</div>
            <div>{readable_arrival}</div>
          </div>
        </div>
      </span>
    </Item>
  );
};

export default FlightItem;