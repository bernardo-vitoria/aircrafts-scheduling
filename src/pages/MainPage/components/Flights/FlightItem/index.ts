import FlightItem from "./FlightItem";

export { FlightItem };