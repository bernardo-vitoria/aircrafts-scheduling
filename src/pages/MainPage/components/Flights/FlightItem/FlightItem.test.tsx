import FlightItem from "./FlightItem";
import { render } from "../../../../../utils/test";
import { flight } from "../../../fixtures";

describe("FlightItem", () => {
  it("should match snapshot", () => {
    const { container } = render(<FlightItem flight={flight} />);

    expect(container).toMatchSnapshot();
  });
});
