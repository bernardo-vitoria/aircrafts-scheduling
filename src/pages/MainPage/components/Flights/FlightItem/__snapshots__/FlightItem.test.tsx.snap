// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlightItem should match snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-poj73t-MuiPaper-root"
  >
    <span
      class="flight-item"
    >
      <div>
        <b>
          AS1001
        </b>
      </div>
      <div
        class="card-body"
      >
        <div
          class="card-body_column"
        >
          <div>
            LFSB
          </div>
          <div>
            06:00
          </div>
        </div>
        <div
          class="card-body_column"
        >
          <div>
            LFMN
          </div>
          <div>
            07:15
          </div>
        </div>
      </div>
    </span>
  </div>
</div>
`;
